### PR TITLE
Filter alerts

### PR DESCRIPTION
**Pull request by shokada**

### CODE REVIEW

File: promgen/tasks.py [hunk 1]
```diff
@@ -12,6 +13,12 @@
 
 logger = logging.getLogger(__name__)
 
+@shared_task
+def index_alert(alert_pk):
+    alert = models.Alert.objects.get(pk=alert_pk)
+    labels = json.loads(alert.body)["commonLabels"]
```
Comment (kfdm) [inline promgen/tasks.py]: I think this could be written as `labels = alert.json.get('commonLabels')` since Alert model defines a `.json` property

Comment (shokada) [reply promgen/tasks.py]: Fixed

File: promgen/templates/promgen/alert_list.html [hunk 2]
```diff
@@ -4,7 +4,12 @@
 
 {% breadcrumb label="Alerts" %}
 
-{% include 'promgen/pagination_short.html' %}
+<form class="navbar-form navbar-left" action="/alert">
+    <div class="form-group">
+        <input name="search" type="text" value="{{ search }}" class="form-control" placeholder="Service, Project">
```
Comment (kfdm) [inline promgen/templates/promgen/alert_list.html]: I think you can write this as `value="{{ request.GET.search }}"`  and then you do not have to implement `get_context_data` in the view

File: promgen/views.py [hunk 3]
```diff
@@ -1019,13 +1019,28 @@ def post(self, request, *args, **kwargs):
...
         tasks.process_alert.delay(alert.pk)
+        tasks.index_alert.delay(alert.pk)
         return HttpResponse("OK", status=202)
 
 
 class AlertList(LoginRequiredMixin, ListView):
     paginate_by = 20
-    queryset = models.Alert.objects.order_by('-created')
 
+    def get_context_data(self, **kwargs):
```
Comment (kfdm) [inline promgen/views.py]: I think I would probably leave the queryset at the top, and then implement it something like 

```python
class AlertList(LoginRequiredMixin, ListView):
    paginate_by = 20
    queryset = models.Alert.objects.order_by("-created")

    def get_queryset(self):
        search = self.request.GET.get("search")
        if search:
            return self.queryset.filter(
                Q(alertlabel__name="Service", alertlabel__value__icontains=search)
                | Q(alertlabel__name="Project", alertlabel__value__icontains=search)
            )
        return self.queryset
```

**Comment by kfdm** [comment]
May also be worthwhile to implement a management command [1] to allow to queue alerts to be re-indexed.

```python
for alert in Alert.objects.all():
    task. index_alert.delay(alert.pk)
```


[1]: https://docs.djangoproject.com/ja/2.2/howto/custom-management-commands/

File: promgen/views.py (hunk 3 above)
Comment (shokada) [reply promgen/views.py]: fixed

File: promgen/templates/promgen/alert_list.html (hunk 2 above)
Comment (shokada) [reply promgen/templates/promgen/alert_list.html]: fixed

**Comment by shokada** [comment]
> May also be worthwhile to implement a management command 1 to allow to queue alerts to be re-indexed.
I will implement that.

**Comment by shokada** [comment]
@kfdm How about this? https://github.com/line/promgen/pull/242/commits/9a1307e06f0a3a96bb478cf26af78ef1c17be5ca

File: promgen/views.py [hunk 4]
```diff
@@ -1033,8 +1033,15 @@ def get_queryset(self):
             return self.queryset.filter(
                 Q(alertlabel__name="Service", alertlabel__value__contains=search)
                 | Q(alertlabel__name="Project", alertlabel__value__contains=search)
+                | Q(alertlabel__name="Job", alertlabel__value__contains=search)
             )
-        return self.queryset
+
+        qs = self.queryset
+        for key, value in self.request.GET.items():
+            if key == "page" or key == "search":
```
Comment (kfdm) [inline promgen/views.py]: Could probably write this as `if key in ["page", "search"]`

File: promgen/views.py [hunk 5]
```diff
@@ -1033,8 +1033,15 @@ def get_queryset(self):
             return self.queryset.filter(
                 Q(alertlabel__name="Service", alertlabel__value__contains=search)
                 | Q(alertlabel__name="Project", alertlabel__value__contains=search)
+                | Q(alertlabel__name="Job", alertlabel__value__contains=search)
```
Comment (kfdm) [inline promgen/views.py]: May be better to do `alertlabel__value__icontains` so that upper/lower case do not matter

File: promgen/views.py (hunk 4 above)
Comment (shokada) [reply promgen/views.py]: fixed

File: promgen/views.py (hunk 5 above)
Comment (shokada) [reply promgen/views.py]: fixed

**Pull request merged**